### DISCUSSION TITLE
How to trigger repository_dispatch event for non-default branch?

### DISCUSSION THREAD

**Discussion by phil-opp** [community/community#24657] [Apps, API and Webhooks]
<p>The Github Actions documentation mentions the <code>repository_dispatch</code> event in the <a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/events-that-trigger-workflows#external-events-repository_dispatch" rel="nofollow noopener"><em>External Events</em></a> section. The table in this sections says that the <code>GITHUB_REF</code> points to the <em>“Branch that received dispatch”</em>. My question now is: How can I specify the branch when sending the <code>repository_dispatch</code> event? The <a href="https://developer.github.com/v3/repos/#create-a-repository-dispatch-event" rel="nofollow noopener">API documentation</a> does not mention any branch related input argument.</p>
<p>Thanks!</p>

**Comment by AndreaGriffiths11** [comment]
<p>Hi <a class="mention" href="/phil-opp">@phil-opp</a>,</p>
<p>Thank you for being here! I wish I had better news for you, the <code>repository_dispatch</code> only triggers on the master branch, and it is not possible to specify a branch. We have received several pieces of feedback about this limitation and have an internal issue tracking a feature request, to which I’ve added your feedback. </p>

**Reply by scanaaron** [reply]
We would love it if we could trigger on: schedule: jobs on branches other than our main branch. Thanks!

**Reply by dersimn** [reply]
https://github.com/marketplace/actions/workflow-dispatch

**Comment by phil-opp** [comment]
<p>Thanks for clarifying! Perhaps the documentation at <a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/events-that-trigger-workflows#external-events-repository_dispatch" rel="nofollow noopener">https://help.github.com/en/actions/automating-your-workflow-with-github-actions/events-that-trigger-workflows#external-events-repository_dispatch</a> should be updated to reflect this (the <code>GITHUB_REF</code> column says “Branch that received dispatch” instead of “default branch”).</p>

**Reply by Mayureshgithub** [reply]
I ain't getting proper answer. Could please give one example.

**Comment by 0xGuybrush** [comment]
<p>Also, when the documentation links “Webhook Event Payload: repository_dispatch” through to <a href="https://developer.github.com/v3/activity/events/types/#repositorydispatchevent" rel="nofollow noopener">the repositorydispatchevent type</a> on the v3 version of the documentation, it makes it seem like I can use the payload body there, but trying to use branch gives:</p>
<pre><code>{
  "message": "Invalid request.\n\n\"branch\" is not a permitted key.",
  "documentation_url": "https://developer.github.com/v3"
}
</code></pre>
<p>Both are a bit confusing, would be good to update if only default branch is possible.</p>

**Reply by gdi-automates** [reply]
I get the same issue

**Comment by mimoo** [comment]
<p>My problem with the repository dispatch is that it needs a token with repo permission. Isn’t it possible to severely limit the scope of the token in order to avoid consequences if it gets compromised?</p>

**Comment by phil-opp** [comment]
<p>This seems like a different problem, so it’s probably better to create a new topic for it.</p>

**Comment by tombrus** [comment]
<p>Since we need this a lot…</p>
<p>two questions:</p>
<ul>
<li>any news on this since last November?</li>
<li>could I mimic the branch by passing it’s name in the <strong>client_payload</strong> and switching to that branch in my action script? (will this build be registered by github as a build on  <strong>master</strong> or on the requested branch?)</li>
<li>is there any other webhook I could (mis)-use to signal a rebuild on a certain branch?</li>
</ul>
<p>Our situation is as follows:</p>
<ul>
<li>repository “lib” contains our library and publishes a package</li>
<li>repository “app” uses the published package to make our app</li>
<li>we use synchronous branches on both repositories</li>
<li>regularly we push to both repos at the same time after bumping the version number of the package (in lib) and its use (in app).</li>
<li>because they both start building at the same time the app will not find the new version of the lib and it fails.</li>
<li>we would like the lib to re-trigger the app as soon as a new lib-package is published, which means the second build of app would find the indicated version of the lib and will succeed</li>
<li>currently we have to manually baby-sit the app-build to start after lib was published</li>
<li>if lib could directly trigger a build of app on whatever branch it is publishing we would automate this.</li>
</ul>
<p>This is a simplified explaination, in reality we now have 4 repos that inter-depend and that number will probably grow. This will make the baby-sitting very time-consuming.</p>

**Comment by missedone** [comment]
<p>is there any update on this issue? it makes the repository_dispatch useless if only triggered on master (or default) branch</p>

**Comment by Evilweed** [comment]
<p><a class="mention" href="/AndreaGriffiths11">@AndreaGriffiths11</a> Any update on this?</p>
<p>This is available in GitLab and allows creating powerful workflows and easy iteration on them (as you don’t have to commit temp workflows to master while developing on feature branch)</p>

**Comment by francisfuzz** [comment]
<h3>
<a name="clarification-which-branch-a-repository-dispatch-is-triggered-for-memo-1" class="anchor" href="#clarification-which-branch-a-repository-dispatch-is-triggered-for-memo-1"></a>Clarification: which branch a repository dispatch is triggered for :memo:
</h3>
<aside class="quote no-group" data-username="phil-opp" data-post="3" data-topic="14470">
<div class="title">
<div class="quote-controls"></div>
 phil-opp:</div>
<blockquote>
<p>Perhaps the documentation at <a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/events-that-trigger-workflows#external-events-repository_dispatch">https://help.github.com/en/actions/automating-your-workflow-with-github-actions/events-that-trigger-workflows#external-events-repository_dispatch </a> should be updated to reflect this (the <code>GITHUB_REF</code> column says “Branch that received dispatch” instead of “default branch”).</p>
</blockquote>
</aside>
<aside class="quote no-group" data-username="another-dave" data-post="4" data-topic="14470">
<div class="title">
<div class="quote-controls"></div>
 another-dave:</div>
<blockquote>
<p>Also, when the documentation links “Webhook Event Payload: repository_dispatch” through to <a href="https://developer.github.com/v3/activity/events/types/#repositorydispatchevent">the repositorydispatchevent type </a> on the v3 version of the documentation, it makes it seem like I can use the payload body there</p>
</blockquote>
</aside>
<p>That’s great feedback. Our team recently updated the documentation to include a <code>note</code> below <a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#repository_dispatch">the <code>repository_dispatch</code> documentation</a>:</p>
<blockquote>
<p><strong>Note:</strong>  This event will only trigger a workflow run if the workflow file is on the  <code>master</code>  or default branch.</p>
</blockquote>
<hr>
<h3>
<a name="workaround-specifying-a-branch-using-the-repository-dispatch-event-construction_worker_man-2" class="anchor" href="#workaround-specifying-a-branch-using-the-repository-dispatch-event-construction_worker_man-2"></a>Workaround: specifying a branch using the repository dispatch event :construction_worker_man:
</h3>
<aside class="quote no-group quote-modified" data-username="tombrus" data-post="7" data-topic="14470">
<div class="title">
<div class="quote-controls"></div>
 tombrus:</div>
<blockquote>
<ul>
<li>could I mimic the branch by passing it’s name in the <strong>client_payload</strong> and switching to that branch in my action script? (will this build be registered by github as a build on <strong>master</strong> or on the requested branch?)</li>
</ul>
</blockquote>
</aside>
<p>Here’s an example repository that showcases using the repository dispatch event and <code>actions/checkout@v2</code> specifying a <code>branch</code> in the <code>client_payload</code> field:</p>
<aside class="onebox allowlistedgeneric" data-onebox-src="https://github.com/actions-packages-examples/branch-builds">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/actions-packages-examples/branch-builds" target="_blank" rel="noopener">GitHub</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/15894826/181085882-5e803e82-fcaa-476c-b955-2f73c07fc510.png" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/3/b/3bab8cc24f981d669d803aca5df6af8cc5a89d37_2_690x345.png, https://aws1.discourse-cdn.com/github/optimized/3X/3/b/3bab8cc24f981d669d803aca5df6af8cc5a89d37_2_1035x517.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/3/b/3bab8cc24f981d669d803aca5df6af8cc5a89d37.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/3/b/3bab8cc24f981d669d803aca5df6af8cc5a89d37_2_10x10.png"></div>

<h3><a href="https://github.com/actions-packages-examples/branch-builds" target="_blank" rel="noopener">GitHub - actions-packages-examples/branch-builds: Example for demonstrating...</a></h3>

  <p>Example for demonstrating branch builds using the repository dispatch event and actions/checkout@v2. - GitHub - actions-packages-examples/branch-builds: Example for demonstrating branch builds usin...</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>In the workflow file, it’s possible to leverage that input to checkout a specific branch (like what’s <a href="https://github.com/actions-packages-examples/branch-builds/blob/f92a17dcabf3757cd51c40d7c4b1c18094b7de6c/.github/workflows/main.yml#L14-L16">defined on these lines of the workflow file</a>).</p>
<p>However, the <code>Actions</code> tab will associate the workflow run triggered by a repository dispatch to the default branch (in this case, <code>master</code>) rather than the specified branch <a href="https://github.com/actions-packages-examples/branch-builds/actions?query=workflow%3AMain+event%3Arepository_dispatch">as evidenced in this query</a>.</p>
<hr>
<h3>
<a name="update-the-latest-news-on-the-repository-dispatch-feature-newspaper-3" class="anchor" href="#update-the-latest-news-on-the-repository-dispatch-feature-newspaper-3"></a>Update: the latest news on the repository dispatch feature :newspaper:
</h3>
<aside class="quote no-group" data-username="tombrus" data-post="7" data-topic="14470">
<div class="title">
<div class="quote-controls"></div>
 tombrus:</div>
<blockquote>
<ul>
<li>any news on this since last November?</li>
</ul>
</blockquote>
</aside>
<aside class="quote no-group" data-username="missedone" data-post="8" data-topic="14470">
<div class="title">
<div class="quote-controls"></div>
 missedone:</div>
<blockquote>
<p>is there any update on this issue?</p>
</blockquote>
</aside>
<p>:wave: Hello! Neither <a class="mention" href="/AndreaGriffiths11">@AndreaGriffiths11</a> or I are able to speak to upcoming product updates. However, any new ships are either published in the form of a <a href="https://github.blog/changelog/">changelog entry</a> or an article in <a href="https://github.blog/">The GitHub Blog</a>.</p>
<p>Today, the best place to share requests for new or updated features is <a href="https://support.github.com/contact/feedback">through our official product feedback form</a>. Our product team working on Actions monitors that channels and considers each entry, but they may not be able to respond to every submission.</p>

**Comment by phil-opp** [comment]
<p>Thanks for your reply!</p>
<p>Unfortunately, the workaround you mentioned has a big drawback: It runs the workflow file specified on the default branch, not the workflow file of the checked out branch. So this only really works if the two branches have the same build/test instructions.</p>
<p>Some examples where this doesn’t work:</p>
<ul>
<li>If we have a <code>docs</code> branch with a workflow file that converts some markdown documentation to HTML. Using your workaround, GitHub Actions would try to run the normal build instructions of the default branch (e.g. compile some code) on the markdown sources, which of course doesn’t work.
<ul>
<li>To prevent this issue, you could copy the markdown build instructions into the workflow file of the default branch. However, now you have to manually keep the two workflow files in sync when the markdown build instructions change at some point (and you need to remember to do it). This also means that any change to the markdown build instructions has to be coordinated across multiple branches, which will make things much more complicated.</li>
</ul>
</li>
<li>Even with branches that have similar build instructions, it is very easy to get things wrong with that workaround. For example, consider some build instructions that use the current branch name for conditionally skipping checks. Another risk is that you probably won’t notice if there were slight differences in the two workflow files, which might result in inconsistent build results depending on how the build was triggered. It could even lead to corrupted release artifacts if the workaround is used to trigger a release of a different branch with slightly different build instructions.</li>
</ul>
<p>So this workaround only works in the very specific case where both branches have exactly the same build instructions and use no advanced features such as skipping steps based on the current branch name.</p>
<p>To solve this problem for more use case, we need some way to trigger a workflow run on a different branch <em>using the workflow file of that branch</em>. I’ll send a feature request for that through the form you linked and I encourage others to do so too.</p>

**Comment by Nitive** [comment, marked as answer]
<p>A new <code>workflow_dispatch</code> event can be used to trigger workflow for non-default branch.<br>
<a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#workflow_dispatch" class="onebox" target="_blank" rel="nofollow noopener">https://docs.github.com/en/actions/reference/events-that-trigger-workflows#workflow_dispatch</a></p>

**Reply by jamesmortensen** [reply]
This answer gave me an idea.  

Since workflow_dispatch does allow triggering workflows on different branches, could you use Puppeteer in to go to github.com/org/repo/actions/workflow/test.yml, find the workflow, locate the workflow dispatch button with:

`document.querySelector("#repo-content-pjax-container > split-page-layout > div > div > div.PageLayout-region.PageLayout-content > div > div > div.Box.Box--responsive.mt-3 > div.flash.flash-full > details > summary").click()`

and then expand the drop down with this:

`document.querySelector("#repo-content-pjax-container > split-page-layout > div > div > div.PageLayout-region.PageLayout-content > div > div > div.Box.Box--responsive.mt-3 > div.flash.flash-full > details > div > div > div > form:nth-child(1) > div.branch-selection.mb-2 > details > summary > span.css-truncate-target").click()`

select the branch with this:

```
BRANCH_NAME='your-branch-here';
var iter = document.evaluate("//button/span[contains(., '"+BRANCH_NAME+"')]", document, null, XPathResult.ANY_TYPE, null );
var elem = iter.iterateNext();
elem.click();
```

And then finally click the "Run Workflow" button:

```
document.querySelector("#repo-content-pjax-container > split-page-layout > div > div > div.PageLayout-region.PageLayout-content > div > div > div.Box.Box--responsive.mt-3 > div.flash.flash-full > details > div > div > div > form:nth-child(2) > button").click();
```

The tricky part might be logging into GitHub from within the workflow.  It could be done with a bot user which has no two-factor auth enabled and would rely on GitHub not putting up any "are you a robot" blockers.

**Comment by francisfuzz** [comment]
<p><a class="mention" href="/nitive">@nitive</a> - Ah, thanks for mentioning that! Our team also published a <a href="https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/">changelog entry for this feature</a> that might be worth checking out as well. :v:</p>

**Comment by crazy-matt** [comment]
<aside class="quote no-group" data-username="Evilweed" data-post="9" data-topic="14470">
<div class="title">
<div class="quote-controls"></div>
 Evilweed:</div>
<blockquote>
<p>This is available in GitLab and allows creating powerful workflows and easy iteration on them (as you don’t have to commit temp workflows to master while developing on feature branch)</p>
</blockquote>
</aside>
<p>That’s in a simple single sentence exactly why a ci/cd tool would need this capability… Please Github…</p>
<aside class="quote group-github-employees" data-username="francisfuzz" data-post="10" data-topic="14470">
<div class="title">
<div class="quote-controls"></div>
 francisfuzz:</div>
<blockquote>
<p>In the workflow file, it’s possible to leverage that input to checkout a specific branch (like what’s <a href="https://github.com/actions-packages-examples/branch-builds/blob/f92a17dcabf3757cd51c40d7c4b1c18094b7de6c/.github/workflows/main.yml#L14-L16" rel="noopener nofollow ugc">defined on these lines of the workflow file</a>).</p>
</blockquote>
</aside>
<p>But this alternative could make it. It would make sense then for Github to display the checkout ref on the Actions tab to give more insight about what happened on our runs.</p>

**Comment by jgehrcke** [comment]
<p><a class="mention" href="/nitive">@nitive</a></p>
<blockquote>
<p>A new <code>workflow_dispatch</code> event can be used to trigger workflow for non-default branch.</p>
</blockquote>
<p>Are you sure that this can address the problem discussed here? I have tried this and failed. Here is what I did:</p>
<ul>
<li>I added a new file <code>.github/workflows/update.yml</code> on a non-default branch, called <code>workflowtest</code>.</li>
<li>I pushed that branch to my repository.</li>
</ul>
<p>I followed <a href="https://docs.github.com/en/free-pro-team@latest/rest/reference/actions#create-a-workflow-dispatch-event" class="inline-onebox" rel="noopener nofollow ugc">Actions - GitHub Docs</a> and tried</p>
<pre><code class="lang-auto">$ curl -H "Authorization: token ${GITHUB_API_TOKEN}" https://api.github.com/repos/jgehrcke/foo/actions/workflows/update.yml/dispatches -d '{"ref":"workflowtest"}'
{
  "message": "Not Found",
  "documentation_url": "https://docs.github.com/rest/reference/actions#create-a-workflow-dispatch-event"
}
</code></pre>
<p>It seems like <code>update.yml</code> isn’t <em>known</em> to the system yet, because it’s <em>not</em> on the default branch.</p>
<p>The <code>Actions</code> tab in the UI of course also doesn’t show the workflow yet.</p>
<p><a class="mention" href="/phil-opp">@phil-opp</a> has tried hard to clarify the challenge he’d like to overcome:</p>
<blockquote>
<p>we need some way to trigger a workflow run on a different branch <em>using the workflow file of that branch</em></p>
</blockquote>
<p>This is really important for the developer experience around building workflows/actions. I’ve commented on that before, here: <a href="https://github.community/t/how-to-develop-iterate-on-debug-a-workflow-file-in-a-non-default-branch/17166/3" class="inline-onebox">How to develop (iterate on, debug!) a workflow file in a non-default branch? - #3 by brightran</a></p>
<p>I don’t want to commit to main/master before having tested the workflow properly.</p>

**Comment by Nitive** [comment]
<p><a class="mention" href="/jgehrcke">@jgehrcke</a> we found a workaround to trigger <code>workflow_dispatch</code> workflow committed to non-default branch.</p>
<ol>
<li>Create a workflow with <code>push</code> trigger like this one</li>
</ol>
<pre><code class="lang-yaml">name: My workflow
on:
  push:

jobs:
  build-tools:
    runs-on: ubuntu-latest
    steps:
      - run: echo 1
</code></pre>
<ol start="2">
<li>Push it</li>
<li>Change <code>push</code> to <code>workflow_dispatch</code> commit and push the change</li>
<li>Change the workflow whatever you like</li>
<li>Trigger workflow from API</li>
</ol>
<p>We are not happy that we have to make an extra commit every time, but at least we don’t have to commit workflow to default branch</p>

**Comment by HarikrishnanBalagopal** [comment]
<p>I am having the same issue but in my case the workflow yaml is on the default branch <code>main</code></p>
<p>I can dispatch the workflow if I specify <code>ref</code> as <code>main</code> or <code>refs/heads/main</code><br>
but it fails if I try any other branch</p>
<pre><code class="lang-auto">$ curl   -X POST   -H "Accept: application/vnd.github.v3+json" -H 'Authorization: token ***'   https://api.github.com/repos/HarikrishnanBalagopal/yohoo/actions/workflows/fooworkflow.yml/dispatches   -d '{"ref":"refs/heads/foo"}'
{
  "message": "Workflow does not have 'workflow_dispatch' trigger",
  "documentation_url": "https://docs.github.com/rest/reference/actions#create-a-workflow-dispatch-event"
}
</code></pre>
<p>To be clear this error is different from the error for non existent workflows:</p>
<pre><code class="lang-auto">$ curl   -X POST   -H "Accept: application/vnd.github.v3+json" -H 'Authorization: token ***'   https://api.github.com/repos/HarikrishnanBalagopal/yohoo/actions/workflows/fooworkflowsss.yml/dispatches   -d '{"ref":"refs/heads/hmmm"}'
{
  "message": "Not Found",
  "documentation_url": "https://docs.github.com/rest/reference/actions#create-a-workflow-dispatch-event"
}
</code></pre>
<p>and this error is also different from the error for non-existent branches:</p>
<pre><code class="lang-auto">$ curl   -X POST   -H "Accept: application/vnd.github.v3+json" -H 'Authorization: token ***'   https://api.github.com/repos/HarikrishnanBalagopal/yohoo/actions/workflows/fooworkflow.yml/dispatches   -d '{"ref":"refs/heads/weird"}'
{
  "message": "No ref found for: refs/heads/weird",
  "documentation_url": "https://docs.github.com/rest/reference/actions#create-a-workflow-dispatch-event"
}
</code></pre>

**Comment by HarikrishnanBalagopal** [comment]
<p>I got it working after committing the workflow yaml to both the default branch and the branch you want to run on.<br>
So in my case had to commit to both <code>main</code> and <code>foo</code></p>

**Comment by kaihendry** [comment]
<p>Whoa, I hope Github developers make this less of a mess.</p>
<ol>
<li><a href="https://github.com/kaihendry/count/blob/c1ade5ccad4d348658a39e02d6144ebe209b853f/.github/workflows/serverless.yml" class="inline-onebox" rel="noopener nofollow ugc">count/serverless.yml at c1ade5ccad4d348658a39e02d6144ebe209b853f · kaihendry/count · GitHub</a></li>
<li><a href="https://github.com/kaihendry/count/blob/e9d58727e08af4bb6a9a7003481b088f56be1fe7/.github/workflows/serverless.yml" class="inline-onebox" rel="noopener nofollow ugc">count/serverless.yml at e9d58727e08af4bb6a9a7003481b088f56be1fe7 · kaihendry/count · GitHub</a></li>
</ol>
<p>Say I modify 2, i.e. the workflow files are not synced. Which one gets run? It appear 2 gets run. So I’m thinking 1. could be a LOT simpler.</p>
<aside class="onebox githubfolder">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/kaihendry/count/actions" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:400/400;"><img src="https://avatars0.githubusercontent.com/u/765871?s=400&amp;amp;v=4" class="thumbnail" width="400" height="400"></div>

<h3><a href="https://github.com/kaihendry/count/actions" target="_blank" rel="noopener nofollow ugc">kaihendry/count</a></h3>

<p><a href="https://github.com/kaihendry/count/actions" target="_blank" rel="noopener nofollow ugc">ns</a></p>

  <p><span class="label1">Comparing serverless platforms. Contribute to kaihendry/count development by creating an account on GitHub.</span></p>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by AZ-X** [comment]
<p>Any updates? I just want to get a workaround if the parameters could pass through an overridden value of  GITHUB_SHA.<br>
For example, git create ref/tag API</p><aside class="onebox githubblob">
  <header class="source">
      <a href="https://github.com/marvinpinto/actions/blob/4d6da7f4aa2ff406ad115549df3d9972487ce797/packages/automatic-releases/src/main.ts#L295" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>
  <article class="onebox-body">
    <h4><a href="https://github.com/marvinpinto/actions/blob/4d6da7f4aa2ff406ad115549df3d9972487ce797/packages/automatic-releases/src/main.ts#L295" target="_blank" rel="noopener nofollow ugc">marvinpinto/actions/blob/4d6da7f4aa2ff406ad115549df3d9972487ce797/packages/automatic-releases/src/main.ts#L295</a></h4>
<pre class="onebox"><code class="lang-ts"><ol class="start lines" start="285" style="counter-reset: li-counter 284 ;">
<li>  context.sha,</li>
<li>);</li>
<li>
<li>const changelog = await getChangelog(client, context.repo.owner, context.repo.repo, commitsSinceRelease);</li>
<li>
<li>if (args.automaticReleaseTag) {</li>
<li>  await createReleaseTag(client, {</li>
<li>    owner: context.repo.owner,</li>
<li>    ref: `refs/tags/${args.automaticReleaseTag}`,</li>
<li>    repo: context.repo.repo,</li>
<li class="selected">    sha: context.sha,</li>
<li>  });</li>
<li>
<li>  await deletePreviousGitHubRelease(client, {</li>
<li>    owner: context.repo.owner,</li>
<li>    repo: context.repo.repo,</li>
<li>    tag: args.automaticReleaseTag,</li>
<li>  });</li>
<li>}</li>
<li>
<li>const releaseUploadUrl = await generateNewGitHubRelease(client, {</li>
</ol></code></pre>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by ghost** [comment]
Any update on this?
We too are looking to trigger repository_dispatch on different branches

**Comment by zhangguanzhang** [comment]
same issue on 
```
curl \
-X POST https://api.github.com/repos/zhangguanzhang/xxxx/dispatches \
-H "Accept: application/vnd.github.everest-preview+json" \
-H "Authorization: token xxxx" \
-d '{
    "event_type":"xxxxxx",
    "client_payload":{
        "target":"dev"
    }
}'
```

**Comment by jamesmortensen** [comment]
I think there's some confusion in this thread.  In the GitHub Action docs, I thought the feature everyone is looking for is called "repository_dispatch" but in the REST documentation, it's referred to as "workflow_dispatch", which in GitHub Actions is a manually triggered event from the UI.

https://docs.github.com/en/rest/actions/workflows#create-a-workflow-dispatch-event

https://docs.github.com/en/rest/repos/repos#create-a-repository-dispatch-event

In the docs, I can see that both can be triggered via a REST call, and there are slight differences in the request paths and data payloads.  Would someone at GitHub be able to clarify the use-cases for both of these in the docs?  What are the advantages of one over the other?  The title of the post is about repository_dispatch, but people are also asking about and suggesting workflow_dispatch, which might be leading to some confusion.

Thank you!
James

**Comment by jaderyang-sms** [comment]
Any update on this?  or any workaround?

**Reply by jamesmortensen** [reply]
If I were to get really crazy about this, I would create a workflow on the default branch that gets triggered by repository dispatch. As part of the payload, I might include the branch name, or just hard code it.

In that workflow, I would then clone the repo, change branches to the target branch, and then make an empty commit and push it to that branch, triggering the target workflow by a push.

The end result is this: We kick-off the process with repository dispatch and eventually trigger a workflow which runs on a branch, triggered via proxy.

You can do an empty commit like this:
`git commit --allow-empty -m "run workflow" `

**Comment by amit-teachmint** [comment]
Hi All, 
I find this working after specifying branch name in client payload

`curl --request POST  -H "Accept: application/vnd.github+json"  -H "X-GitHub-Api-Version: 2022-11-28"  --url 'https://api.github.com/repos/OWNER/REPO/dispatches'  --header 'authorization: Bearer <TOKEN>'  --data '{"event_type": "NAME","client_payload":{"ref":"BRANCH","INPUTS_KEY":"INPUTS_VALUE"}}' `

**Reply by racitores** [reply]
Agree. I think the best alternative when using the API is to include the branch in within the payload as this [doc](https://docs.github.com/en/actions/managing-workflow-runs/manually-running-a-workflow#running-a-workflow-using-the-rest-api) states.

**Reply by standage-thanh** [reply]
Yes, we can do with [workflow-dispatch-event](https://docs.github.com/en/rest/actions/workflows?apiVersion=2022-11-28#create-a-workflow-dispatch-event)

**Reply by ppagnoni** [reply]
It doesn't work for me!
It always calls the master workflow 😟 

```
  -d '{
    "event_type": "et_result",
    "client_payload": {
      "message": "Test msg",
      "ref": "feature/repository-dispatch-test"
    }
  }'
```

**Comment by wparad** [comment]
Just got so annoyed that this doesn't work out of the box for pull requests, that I just had to build it myself: https://github.com/marketplace/actions/rebuild-stale-pullrequests
